// _digo@0.2.0@digo/lib/utility/httpServer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const http = require("http");
const nu = require("url");
/**
 * 表示一个 Http 服务器。
 */
class HttpServer extends events_1.EventEmitter {
    constructor() {
        super(...arguments);
        /**
         * 底层 HTTP 服务器。
         */
        this._server = http.createServer(this.processRequest.bind(this))
            .on("listening", this.onStart.bind(this))
            .on("close", this.onStop.bind(this))
            .on("error", this.onError.bind(this));
        /**
         * 当前服务器的虚拟路径。
         */
        this.virtualPath = "/";
    }
    /**
     * 判断当前服务器是否正在监听。
     */
    get isListening() { return this._server.listening; }
    /**
     * 获取当前服务器的请求数。
     */
    get connectionCount() { return this._server.connections; }
    /**
     * 当前服务器的超时毫秒数。默认为 120000。
     */
    get timeout() { return this._server.timeout; }
    set timeout(value) { this._server.timeout = value; }
    /**
     * 获取当前服务器的主页地址。
     */
    get url() {
        const addr = this._server.address() || { address: "0.0.0.0", port: 80 };
        return `http://${addr.address === "0.0.0.0" ? "localhost" : addr.address}${addr.port === 80 ? "" : ":" + addr.port}${this.virtualPath}`;
    }
    /**
     * 当服务器启动时回调。
     */
    onStart() {
        this.emit("start");
    }
    /**
     * 当服务器停止时回调。
     */
    onStop() {
        this.emit("stop");
    }
    /**
     * 当服务器错误时执行。
     * @param e 当前发生的错误。
     */
    onError(e) {
        this.emit("error", e);
    }
    /**
     * 当被子类重写时负责处理所有请求。
     * @param req 当前的请求对象。
     * @param res 当前的响应对象。
     */
    processRequest(req, res) {
        this.emit("request", req, res);
    }
    /**
     * 监听指定的地址。
     * @param url 要监听的地址。
     * @param callback 启动完成的回调函数。
     */
    listen(url = "http://0.0.0.0:0/", callback) {
        const urlObject = nu.parse(url);
        this.virtualPath = urlObject.path || this.virtualPath;
        this._server.listen(+urlObject.port || 0, urlObject.hostname, callback);
    }
    /**
     * 关闭当前服务器。
     * @param callback 关闭的回调函数。
     */
    close(callback) {
        this._server.close(callback);
    }
}
exports.HttpServer = HttpServer;
exports.default = HttpServer;
